<a href="https://colab.research.google.com/github/gcastiblancoc/AlgorithmsUN2024II/blob/main/Lab3/gcastiblancoc_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2200.180
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4400.36
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

^C


### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py


^C


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [ ]:
%%writefile output0.txt
0

Writing output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
5
700000000 900000000 1200000000 1500000000 1800000000

Writing input4.txt


In [ ]:
1500000000 * 1800000000

2700000000000000000

In [ ]:
%%writefile output4.txt
2700000000000000000

Writing output4.txt


In [ ]:
%%writefile input5.txt
3
-5 10 15

Writing input5.txt


In [ ]:
%%writefile output5.txt
6

Writing output5.txt


In [ ]:
%%writefile input6.txt
6
5000 60 7 3500 10 90

Writing input6.txt


In [ ]:
5000 * 3500

17500000

In [ ]:
%%writefile output6.txt
17500000

Writing output6.txt


In [ ]:
%%writefile input7.txt
6
-4500 -5600 -4300 -300000 -2700 -3200

Writing input7.txt


In [ ]:
-5600 * -300000

1680000000

In [ ]:
%%writefile output7.txt
1680000000

Writing output7.txt


In [ ]:
%%writefile hello.cpp
#include <iostream>
using namespace std;

int main() {
    long long a = -200000;
    long long b = -220000;
    cout << a*b << endl;
    return 0;
}

Writing hello.cpp


In [ ]:
!g++ hello.cpp -o hello
!./hello

44000000000


In [ ]:
%%writefile input8.txt
6
0 0 0 0 0 7

Writing input8.txt


In [ ]:
0*7

0

In [ ]:
%%writefile output8.txt
0

Writing output8.txt


In [ ]:
import random
random_list = []
for i in range(5):
  random_list.append(int(random.random() * 1000000))
random_list

[882055, 560753, 261947, 858287, 96552]

In [ ]:
%%writefile input9.txt
5
882055 560753 261947 858287 96552

Writing input9.txt


In [ ]:
858287*96552

82869326424

In [ ]:
%%writefile output9.txt
82869326424

Writing output9.txt


In [ ]:
%%writefile input10.txt
4
-8 -12 -15 2

Writing input10.txt


In [ ]:
-12 * -15

180

In [ ]:
%%writefile output10.txt
180

Writing output10.txt


In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {4..10}
do
    ./max_pairwise_product_long_full < input${i}.txt > ctest${i}.txt
    grep -xvFf output${i}.txt ctest${i}.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

150
757056339785


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

^C


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line
    execints++;

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 571 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

571


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

601



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

572


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

^C


In [ ]:
# Estimated time by instruction

295/200003177092


1.4749765693187072e-09

In [ ]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m1.303s
user	0m1.236s
sys	0m0.002s


## Python

In [ ]:
%%writefile max_pairwise_product.py
def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]

                if aux_product > max_product:
                  max_product = aux_product

    return max_product

n = int(input())

numbers = [int(i) for i in input().split()]

print(max_pairwise_product(numbers))

Writing max_pairwise_product.py


In [ ]:
!time python3 max_pairwise_product.py < input11.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product.py", line 19, in <module>
    print(max_pairwise_product(numbers))
  File "/content/max_pairwise_product.py", line 10, in max_pairwise_product
    if aux_product > max_product:
KeyboardInterrupt
^C


In [ ]:
# Verify
!time python3 max_pairwise_product.py < input13.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product.py", line 19, in <module>
    print(max_pairwise_product(numbers))
  File "/content/max_pairwise_product.py", line 10, in max_pairwise_product
    if aux_product > max_product:
KeyboardInterrupt
^C


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


**Stress tests**

In [ ]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
for i in {4..10}
do
    ./max_pairwise_product_long < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

150
757056339785


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 7n + \left[2(n-1)n+1, 2.5(n-1)n\right)= O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 7(10) + 2(9)11   = 279 $$

Calculate estimated time by instruction and number of instructions by second

For the input11.txt

200000 1 2 .. 199999 2000

In [ ]:
%%writefile max_pairwise_product_longexec.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long execints = 0LL;
long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  execints+=1;
    long long aux_product = 0LL;                                             //1
    execints+=1;
    int n = numbers.size();                                                  //1
    execints+=1;
    execints+=1;
    for (int first = 0; first < n; ++first) {                                //2+2n
        execints+=2;
        execints+=1;
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       execints+=2;
             aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
             execints+=1;
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
                execints+=1;
             }
        }
        execints+=1;
    }
    execints+=1;
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;  execints+=1;                                              //1
    std::cin >> n; execints+=1;                                       //1
    std::vector<int> numbers(n);                          //n
    execints+=1;
    execints+=1;
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        execints+=2;
        std::cin >> numbers[i];                           //n
        execints+=1;
    }
    MaxPairwiseProduct(numbers);
    std::cout << "Número de instrucciones: " << execints << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_longexec.cpp


In [ ]:
!g++ max_pairwise_product_longexec.cpp -o max_pairwise_product_longexec
!time ./max_pairwise_product_longexec < input11.txt

^C


In [ ]:
# Estimated time by instruction (rounded values)
155/60003772077

2.583170934672171e-09

In [ ]:
# Estimated number of instructions by second
39999800000/145

275860689.6551724

It is much faster than previous solutions and, therefore, executes more instructions per second.

**Pyhon improved solution 1:**

In [ ]:
%%writefile max_pairwise_product_improved.py
def max_pairwise_product_improved(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product, numbers[first] * numbers[second])

    return max_product

n = int(input())
numbers = [int(i) for i in input().split()]
print(max_pairwise_product_improved(numbers))

Writing max_pairwise_product_improved.py


Test inputs 1 and 2

In [ ]:
%%writefile run_all_test.sh
#comparing all the executions to check the differences between the outputs
for i in {1..2}
do
    python3 max_pairwise_product_improved.py < input$i.txt > pythonOutput$i.txt
    grep -xvFf pythonOutput$i.txt output$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Test and execution time for input 3

In [ ]:
!time python3 max_pairwise_product_improved.py < input3.txt > pythonOutput3.txt
!grep -xvFf pythonOutput3.txt output3.txt


real	0m0.032s
user	0m0.022s
sys	0m0.010s


**20 Stress Tests:**

In [ ]:
import random

def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]

                if aux_product > max_product:
                  max_product = aux_product

    return max_product

def max_pairwise_product_improved(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product, numbers[first] * numbers[second])

    return max_product

for _ in range(20):  # Run the test 20 times
    n = random.randint(2, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {max_pairwise_product_improved(numbers)}")
    print(f"The comparison solution gives the result {max_pairwise_product(numbers)}")
    if max_pairwise_product_improved(numbers) == max_pairwise_product(numbers):
        print("The solution is correct OK")
    else:
        print("ERROR")
    print("-------------------------------------------------")

-------------------------------------------------
Testing with the list: [463958035, 265994366, 753612895, 286433064]
The efficient solution gives the result 349644757914861325
The comparison solution gives the result 349644757914861325
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [797131402, 795681947, 203820138, 769355203, 659796529]
The efficient solution gives the result 634263065958199694
The comparison solution gives the result 634263065958199694
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [875632395, 865411019, 932185468, 248099571, 789465522, 734873492, 387073220, 20701952, 416466403]
The efficient solution gives the result 816251793929035860
The comparison solution gives the result 816251793929035860
The solution is correct OK
-------------------------------------------

Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_improved_count.py
import math
exec_times = 0

def max_pairwise_product_improved(numbers):
    global exec_times
    exec_times += 1
    max_product = 0

    exec_times += 1
    n = len(numbers)

    exec_times += 1
    for first in range(n):
        exec_times += 2

        exec_times += 1
        for second in range(first + 1, n):
            exec_times += 2
            max_product = max(max_product, numbers[first] * numbers[second])
        exec_times += 1
    exec_times += 1

    return max_product

exec_times += 1
n = int(input())

exec_times += n
numbers = [int(i) for i in input().split()]

exec_times += 1
max_pairwise_product_improved(numbers)
print("Number of instructions: ", exec_times)

Writing max_pairwise_product_improved_count.py


In [ ]:
%%writefile run_all_test.sh
for i in {1..3}
do
    echo "Input #$i"
    python3 max_pairwise_product_improved_count.py < input$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
Number of instructions:  146
Input #2
Number of instructions:  146
Input #3
Number of instructions:  66


### Calculate estimated time by instruction and number of instructions by second

Using input 11

In [ ]:
import math

def max_pairwise_product_improved(numbers):
    max_product = 0 # 1
    n = len(numbers) # 1


    for first in range(n): # 2n + 2
        for second in range(first + 1, n): # 2n + (n-1)n
            max_product = max(max_product, numbers[first] * numbers[second]) # n^2

    return max_product

n = int(input()) # 1
numbers = [int(i) for i in input().split()] # n
max_pairwise_product_improved(numbers)


KeyboardInterrupt: Interrupted by user

The number of operations is: $$T (n) = 2n^2 + 4n + 5 = O(n^2)$$

Using input11 (n=20000)

$$T (20000) = 2*20000^2 + 4 * 20000 + 5 ≈ 800080005$$

Calculate time:

In [ ]:
!time python3 max_pairwise_product_improved.py < input11.txt > pythonOutputSort11.txt
!grep -xvFf pythonOutput11.txt output11.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product_improved.py", line 13, in <module>
    print(max_pairwise_product_improved(numbers))
  File "/content/max_pairwise_product_improved.py", line 7, in max_pairwise_product_improved
    max_product = max(max_product, numbers[first] * numbers[second])
KeyboardInterrupt
^C
grep: pythonOutput11.txt: No such file or directory


Python complexity time is estimated the same as the c++ solution. So now let's study it's time execution performance

The code runs for more than 1h: 10min.However, the code did not finish executing, this is because Python is much slower than C++ to carry out the instructions.



For this reason, with input 11, it was not possible to determine the number of instructions per second or the execution delay.

# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$


Python sorting max pairwise product

In [ ]:
%%writefile max_pairwise_product_sort.py
def max_pairwise_product_sort(numbers):
  numbers.sort(reverse=True)
  return numbers[0] * numbers[1]

n = int(input())
numbers = [int(i) for i in input().split()]
print(max_pairwise_product_sort(numbers))

Writing max_pairwise_product_sort.py


Test with inputs 1 and 2



In [ ]:
%%writefile run_all_test.sh
#comparing all the executions to check the differences between the outputs
for i in {1..2}
do
    python3 max_pairwise_product_sort.py < input$i.txt > pythonOutput$i.txt
    grep -xvFf pythonOutput$i.txt output$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Test and time it with input 3

In [ ]:
!time python3 max_pairwise_product_sort.py < input3.txt > pythonOutput3.txt
!grep -xvFf pythonOutput3.txt output3.txt


real	0m0.030s
user	0m0.023s
sys	0m0.008s


20 stress tests

In [ ]:
import random
def max_pairwise_product_sort(numbers):
  numbers.sort(reverse=True)
  return numbers[0] * numbers[1]

for _ in range(20):  # Run the test 20 times
    n = random.randint(2, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {max_pairwise_product_sort(numbers)}")
    print(f"The comparison solution gives the result {max_pairwise_product(numbers)}")
    if max_pairwise_product_sort(numbers) == max_pairwise_product(numbers):
        print("The solution is correct OK")
    else:
        print("ERROR")
    print("-------------------------------------------------")

-------------------------------------------------
Testing with the list: [615569355, 825732558, 47498376, 875424444, 582587335, 799218495, 994296735, 422127153, 351324066, 700449566]
The efficient solution gives the result 870431666408390340
The comparison solution gives the result 870431666408390340
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [32093972, 478220985, 64341158, 263995174, 229952317, 391482358, 16030926, 937772530, 268260803]
The efficient solution gives the result 448462503002542050
The comparison solution gives the result 448462503002542050
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [900179502, 163170836]
The efficient solution gives the result 146883041891403672
The comparison solution gives the result 146883041891403672
The solution is correct OK
-------------

Count executed instructions in code and verify with inputs 1,2 and 3


In [135]:
%%writefile max_pairwise_product_sort_count.py
import math
exec_times = 0
def max_pairwise_product_sort(numbers):
  global exec_times

  exec_times += int(len(numbers)*math.log(len(numbers))) # .sort Time complexity n*log(n)
  numbers.sort(reverse=True)

  exec_times += 1
  return numbers[0] * numbers[1]

exec_times += 1
n = int(input())

exec_times += n
numbers = [int(i) for i in input().split()]

exec_times += 1
max_pairwise_product_sort(numbers)
print("Number of instructions: ", exec_times)

Overwriting max_pairwise_product_sort_count.py


In [136]:
%%writefile run_all_test.sh
for i in {1..3}
do
    echo "Input #$i"
    python3 max_pairwise_product_sort_count.py < input$i.txt
done

Overwriting run_all_test.sh


In [137]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
Number of instructions:  36
Input #2
Number of instructions:  36
Input #3
Number of instructions:  19


Calculate estimated time by instruction and number of instructions by second

Using input11

In [139]:
def max_pairwise_product_sort(numbers):
  numbers.sort(reverse=True) #  [N, N * Log (N)]
  return numbers[0] * numbers[1] #  1

The # of operation is given by .sort python time complexity ( N* log N), and the return statement calculation (1)

$$T (n) = 1 + \left[n, n*log(n)\right] = O(n*log(n))$$

Using input11 (n=20000)

$$T (20000) = 1 + 20000*log(20000) ≈ 86022$$

Calculate time

In [141]:
!time python3 max_pairwise_product_sort.py < input11.txt > pythonOutputSort11.txt
!grep -xvFf pythonOutput11.txt output11.txt


real	0m0.095s
user	0m0.070s
sys	0m0.025s
grep: pythonOutput11.txt: No such file or directory


In [142]:
# Estimated time by instruction
time_per_instruction = 0.170/86022
time_per_instruction

1.9762386366278397e-06

In [143]:
# Estimated number of instructions by second
time_per_second = 86022/0.170
time_per_second

506011.7647058823

### Estimate and verify the maximun n that can be solved in 1 second


$$(1+ n*log(n)) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 38923600$$


C++ sorting max pairwise product

In [144]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_sort(std::vector<int>& numbers) {
    std::sort(numbers.rbegin(), numbers.rend());
    return static_cast<long long>(numbers[0]) * numbers[1];
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << max_pairwise_product_sort(numbers) << "\n";

    return 0;
}

Writing max_pairwise_product_sort.cpp


Tests with inputs 1 and 2

In [148]:
%%writefile run_all_test.sh
#comparing all the executions to check the differences between the outputs
for i in {1..2}
do
    g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
    ./max_pairwise_product_sort < input$i.txt > cppOutput$i.txt
    grep -xvFf cppOutput$i.txt output$i.txt
done

Overwriting run_all_test.sh


In [149]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Test and time it with input 3

In [150]:
!time ./max_pairwise_product_sort < input3.txt > cppOutput3.txt
!grep -xvFf cppOutput3.txt output3.txt


real	0m0.004s
user	0m0.002s
sys	0m0.001s


20 Stress tests

In [151]:
%%writefile stressTestCPlusPlusSort.cpp

#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

long long max_pairwise_product_sort(std::vector<int>& numbers) {
    std::sort(numbers.rbegin(), numbers.rend());
    return static_cast<long long>(numbers[0]) * numbers[1];
}

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());

    for (int i = 0; i < 20; ++i) {
        int n = std::uniform_int_distribution<int>(2, 12)(gen);  // Use a large n
        std::vector<int> numbers(n);
        for (int j = 0; j < n; ++j) {
            numbers[j] = std::uniform_int_distribution<int>(1, 1e9)(gen);  // Generate n random numbers between 1 and 10^9
        }

        std::cout << "-------------------------------------------------\n";
        std::cout << "Testing with the list: ";
        for (int num : numbers) {
            std::cout << num << " ";
        }
        std::cout << "\n";

        long long efficient_result = max_pairwise_product_sort(numbers);
        long long comparison_result = MaxPairwiseProduct(numbers);  // Replace with your comparison function result

        std::cout << "The efficient solution gives the result " << efficient_result << "\n";
        std::cout << "The comparison solution gives the result " << comparison_result << "\n";

        if (efficient_result == comparison_result) {
            std::cout << "The solution is correct OK\n";
        } else {
            std::cout << "ERROR\n";
        }

        std::cout << "-------------------------------------------------\n";
    }

    return 0;
}


Writing stressTestCPlusPlusSort.cpp


In [152]:
!g++ stressTestCPlusPlusSort.cpp -o max_pairwise_product_long
!./max_pairwise_product_long

-------------------------------------------------
Testing with the list: 287663365 578942530 518431249 466062196 773328982 105483189 147975686 
The efficient solution gives the result 447713037361404460
The comparison solution gives the result 447713037361404460
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: 382976973 951829349 821980345 
The efficient solution gives the result 782385016672145405
The comparison solution gives the result 782385016672145405
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: 328647409 373934005 
The efficient solution gives the result 122892441880243045
The comparison solution gives the result 122892441880243045
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the lis

Count executed instructions in code and verify with inputs 1,2 and 3

In [153]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>       /* log10 */

long long execints = 0LL;
int n;

long long max_pairwise_product_sort(std::vector<int>& numbers) {
    execints += n + log10(n);
    std::sort(numbers.rbegin(), numbers.rend());
    execints += 1;
    return static_cast<long long>(numbers[0]) * numbers[1];
}

int main() {
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

     max_pairwise_product_sort(numbers);

    std::cout << execints << "\n";

    return 0;
}

Overwriting max_pairwise_product_sort.cpp


In [154]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
for i in {1..3}
do
    echo "Input #$i"
    ./max_pairwise_product_sort < input$i.txt
done

Overwriting run_all_test.sh


In [155]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
12
Input #2
12
Input #3
7


**Calculate estimated time by instruction and number of instructions by second**

**Using input11**

Calculating operations

In [156]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_sort(std::vector<int>& numbers) {
    std::sort(numbers.rbegin(), numbers.rend()); // N * log(N)
    return static_cast<long long>(numbers[0]) * numbers[1]; // 1
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << max_pairwise_product_sort(numbers) << "\n";

    return 0;
}

Overwriting max_pairwise_product_sort.cpp


The # of operation is given by .sort python time complexity ( N* log N), and the return statement calculation (1)

$$T (n) = 1 + \left[n, n*log(n)\right] = O(n*log(n))$$

Using input11 (n=20000)

$$T (20000) = 1 + 20000*log(20000) ≈ 86022$$

**Calculate time**

In [157]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input11.txt > cppOutput11.txt
!grep -xvFf cppOutput11.txt output11.txt


real	0m0.208s
user	0m0.201s
sys	0m0.003s


**Time by instruction**

In [158]:
time_per_instruction = 0.208/86022
time_per_instruction

2.4179860965799448e-06

Number of instructions by second

In [159]:
time_per_second = 86022/0.208
time_per_second

413567.3076923077

### Estimate and verify the maximun n that can be solved in 1 second


$$(1+ n*log(n)) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 38923600$$

## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

# Python

In [160]:
%%writefile efficientSolution.py
def efficientSolution(numbers):
    largest = 0
    secondLargest = 0
    for i in numbers:
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i

    result= secondLargest * largest
    return result

n = int(input())
numbers = [int(i) for i in input().split()]
print(efficientSolution(numbers))

Writing efficientSolution.py


Test with inputs 1 and 2



In [161]:
!python3 efficientSolution.py < input1.txt
!python3 efficientSolution.py < input2.txt

90
90


Test and time it with input 3

In [162]:
!time python3 efficientSolution.py < input3.txt

40000000000

real	0m0.031s
user	0m0.022s
sys	0m0.008s


20 stress tests

In [163]:
import random
def efficientSolution(numbers):
    largest = 0
    secondLargest = 0
    for i in numbers:
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i

    return secondLargest * largest

def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]
                if aux_product > max_product:
                    max_product = aux_product

    return max_product

for _ in range(20):  # Run the test 10 times
    n = random.randint(1, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {efficientSolution(numbers)}")
    print(f"The comparison solution gives the result {max_pairwise_product(numbers)}")
    if efficientSolution(numbers) == max_pairwise_product(numbers):
        print("The solution is correct OK")
    else:
        print("ERROR")
    print("-------------------------------------------------")




-------------------------------------------------
Testing with the list: [510504663]
The efficient solution gives the result 0
The comparison solution gives the result 0
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [140605486, 930361443, 474621155, 788317279]
The efficient solution gives the result 733420001232273597
The comparison solution gives the result 733420001232273597
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [855134659, 986985468, 119721019, 711213896]
The efficient solution gives the result 844005481616135412
The comparison solution gives the result 844005481616135412
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [10271886, 252604445, 644122967, 997466970, 83451741

Count executed instructions in code and verify with inputs 1,2 and 3


In [164]:
%%writefile efficientSolutionexcec.py
exceintspy= 0
def efficientSolution(numbers):
    global exceintspy
    largest = 0
    exceintspy+=1
    secondLargest = 0
    exceintspy+=1
    exceintspy+=1
    for i in numbers:
        exceintspy+=2
        exceintspy+=1
        if i > largest:
            secondLargest = largest
            exceintspy+=1
            largest = i
            exceintspy+=1
        elif i > secondLargest:
            secondLargest = i
            exceintspy+=1
    exceintspy+=1
    result= secondLargest * largest
    exceintspy+=2 # multiplciation instruction
    return result

n = int(input())
exceintspy+=1
numbers = [int(i) for i in input().split()]
exceintspy+=n
print(efficientSolution(numbers))
exceintspy+=1
print(f"Execution Times: {exceintspy}")

Overwriting efficientSolutionexcec.py


In [165]:
%%writefile run_all_test.sh
for i in {1..3}
do
    echo "Input #$i"
    python3 efficientSolutionexcec.py < input$i.txt
done

Overwriting run_all_test.sh


In [166]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
90
Execution Times: 68
Input #2
90
Execution Times: 51
Input #3
40000000000
Execution Times: 40


Calculate estimated time by instruction and number of instructions by second

In [167]:
!time python3 efficientSolutionexcec.py< input11.txt > ptest.txt
!grep -xvFf output11.txt ptest.txt


real	0m0.169s
user	0m0.145s
sys	0m0.023s
Execution Times: 1200008


In [168]:
# Estimated time by instruction
0.192/1200008

1.599989333404444e-07

In [169]:
# Estimated number of instructions by second
1200008/0.192

6250041.666666667

### Estimate and verify the maximun n that can be solved in 1 second

In [170]:
%%writefile efficientSolution.py
def efficientSolution(numbers):
    largest = 0                     #1
    secondLargest = 0               #1
    for i in numbers:               #2+2n
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i
    #WORST CASE 3n
    result= secondLargest * largest    #3
    return result

n = int(input())                            #1
numbers = [int(i) for i in input().split()] #n
print(efficientSolution(numbers))           #1

Overwriting efficientSolution.py


##The number of instructions executed by the algooirithm with a list of size $n$

$$T(n) = (6n+7) = O(n) $$


By calculating the number of instructions that can be executed on input11 we obtain
$$T(n) = (6(200000)+7) = 1200007 $$

Maximum n that can be solve in 1 s
$$(6n+7) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 113378684$$

# C++

In [172]:
%%writefile efficientSolutionC.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long efficientSolution(std::vector<long long>& numbers) {
    long long largest = 0;
    long long secondLargest = 0;
    for (long long i : numbers) {
        if (i > largest) {
            secondLargest = largest;
            largest = i;
        } else if (i > secondLargest) {
            secondLargest = i;
        }
    }
    return secondLargest * largest;
}

int main() {
    long long n;
    std::cin >> n;
    std::vector<long long> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }
    std::cout << efficientSolution(numbers) << std::endl;
    return 0;
}

Writing efficientSolutionC.cpp


Test with inputs 1 and 2

In [173]:
!g++ efficientSolutionC.cpp -o efficientSolutionC
!./efficientSolutionC< input1.txt
!./efficientSolutionC< input2.txt

90
90


Test and time it with input 3

In [174]:
!g++ efficientSolutionC.cpp -o efficientSolutionC
!time ./efficientSolutionC< input3.txt

40000000000

real	0m0.004s
user	0m0.001s
sys	0m0.002s


**20 Stress tests**

In [175]:
%%writefile stressTest.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib> // Para rand() y srand()
#include <ctime> // Para time()

long long efficientSolution(std::vector<long long>& numbers) {
    long long largest = 0;
    long long secondLargest = 0;
    for (long long i : numbers) {
        if (i > largest) {
            secondLargest = largest;
            largest = i;
        } else if (i > secondLargest) {
            secondLargest = i;
        }
    }
    return secondLargest * largest;
}

long long MaxPairwiseProduct(const std::vector<long long>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}


// Esta es una función de prueba de estrés
void stressTest(int maxN, int maxNumber) {
    srand(time(0)); // Inicializar la semilla del generador de números aleatorios

    for (int test = 0; test < 20; ++test) { // Ejecutar la prueba 20 veces
        int n = rand() % maxN + 1; // Generar un número aleatorio entre 1 y maxN
        std::vector<long long> numbers(n);
        for (int i = 0; i < n; ++i) {
            numbers[i] = rand() % maxNumber + 1; // Generar un número aleatorio entre 1 y maxNumber
        }

        long long efficientResult = efficientSolution(numbers);
        long long comparisonSolution = MaxPairwiseProduct(numbers);
        std::cout << "-------------------------------------------------\n";
        std::cout << "Testing with the list: [" ;
        for (long long number : numbers) {
            std::cout << number << ' ';
        }
        std::cout << "]\n";
        std::cout<<"The efficient solution gives the result " << efficientResult<< '\n';
        std::cout<<"The comparison solution gives the result " << comparisonSolution<< '\n';

        if (efficientResult == comparisonSolution) {
            std::cout << "The solution is correct OK\n";
        } else {
            std::cout << "ERROR\n";
            break;
        }
        std::cout << "-------------------------------------------------\n";
    }
}

int main() {
    stressTest(12, 1000000000); // Ejecutar la prueba de estrés con n hasta 12 y números hasta 10^9
    return 0;
}

Writing stressTest.cpp


In [176]:
!g++ stressTest.cpp -o stressTest
!./stressTest

-------------------------------------------------
Testing with the list: [235092588 527567922 738628108 56678387 52813192 888817021 965137404 172882855 14747342 180742238 936151071 ]
The efficient solution gives the result 903514414416759684
The comparison solution gives the result 903514414416759684
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [506259427 703700130 ]
The efficient solution gives the result 356254824593625510
The comparison solution gives the result 356254824593625510
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [712676118 763346563 73280062 255856334 916014425 ]
The efficient solution gives the result 699236462982171275
The comparison solution gives the result 699236462982171275
The solution is correct OK
-------------------------------------------------
--------

Count executed instructions in code and verify with inputs 1,2 and 3

In [177]:
%%writefile efficientSolutionCexec.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long efficientSolution(std::vector<long long>& numbers, long long& execints) {

    long long largest = 0;
    execints+=1;
    long long secondLargest = 0;
    execints+=1;
    execints+=1;
    for (long long i : numbers) {
        execints+=2;
        execints+=1;
        if (i > largest) {
            secondLargest = largest;
            execints+=1;
            largest = i;
            execints+=1;
        } else if (i > secondLargest) {
            execints+=1;
            secondLargest = i;
            execints+=1;
        }
    }
    execints+=1;
    execints+=1;
    return secondLargest * largest;
}

int main() {
    long long execints =0LL;
    long long n;
    execints+=1;
    std::cin >> n;
    execints+=1;
    std::vector<long long> numbers(n);
    execints+=1;
    execints+=1;
    for (int i = 0; i < n; ++i) {
        execints+=2;
        std::cin >> numbers[i];
        execints+=1;
    }
    efficientSolution(numbers, execints);
    execints+=1;
    execints+=1;
    std::cout << "El numero de instrucciones es: " <<  execints << "\n";
    return 0;
}

Writing efficientSolutionCexec.cpp


In [178]:
%%writefile run_all_test.sh
#!/bin/bash
g++ efficientSolutionCexec.cpp -o efficientSolutionCexec
for i in {1..3}
do
    echo "Input #$i"
    ./efficientSolutionCexec < input$i.txt
done

Overwriting run_all_test.sh


In [179]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
El numero de instrucciones es: 91
Input #2
El numero de instrucciones es: 75
Input #3
El numero de instrucciones es: 57


Calculate estimated time by instruction and number of instructions by second

In [180]:
!g++ efficientSolutionCexec.cpp -o efficientSolutionCexec
!time ./efficientSolutionCexec< input11.txt

El numero de instrucciones es: 1600011

real	0m0.128s
user	0m0.121s
sys	0m0.004s


In [181]:
# Estimated time by instruction
0.071/1600011

4.437469492397239e-08

In [182]:
# Estimated number of instructions by second
1600011/0.071

22535366.197183102

## Estimate and verify the maximun n that can be solved in 1 second

In [183]:
%%writefile efficientSolutionC.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long efficientSolution(std::vector<long long>& numbers) {
    long long largest = 0;             //1
    long long secondLargest = 0;       //1
    for (long long i : numbers) {      //2+2n
        if (i > largest) {
            secondLargest = largest;
            largest = i;
        } else if (i > secondLargest) {
            secondLargest = i;
        }
        //WORST CASE 3n
    }
    return secondLargest * largest; //1
}

int main() {
    long long n; //1
    std::cin >> n; //1
    std::vector<long long> numbers(n); //1
    for (int i = 0; i < n; ++i) {  //2+2n
        std::cin >> numbers[i];   //n
    }
    std::cout << efficientSolution(numbers) << std::endl; //1
    return 0;
}

Overwriting efficientSolutionC.cpp


##The number of instructions executed by the algooirithm with a list of size $n$

$$T(n) = (8n+7) = O(n) $$

By calculating the number of instructions that can be executed on input11 we obtain
$$T(n) = (8(200000)+7) = 1600007 $$

Maximum n that can be solve in 1 s
$$(8n+7) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 85034013$$

## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [184]:
%%writefile max_pairwise_efficient.cpp

#include <iostream>
#include <limits>

long long max_pairwise_efficient() {
    int n;            //1
    std::cin >> n;    //1

    long long biggest_1 = std::numeric_limits<long long>::min(); //1
    long long biggest_2 = std::numeric_limits<long long>::min(); //1

    for (int i = 0; i < n; ++i) {  //2 + 2n
        int new_number;
        std::cin >> new_number;
        if (new_number > biggest_1) {
            biggest_2 = biggest_1;
            biggest_1 = new_number;
        } else if (new_number > biggest_2) {
            biggest_2 = new_number;
        }
    }

    return biggest_1 * biggest_2; //1
}

int main() {
    std::cout << max_pairwise_efficient() << std::endl; //1
    return 0; //1
}




Writing max_pairwise_efficient.cpp


In [185]:
!g++ max_pairwise_efficient.cpp -o max_pairwise_efficient -std=c++11
!./max_pairwise_efficient

^C


Test with inputs 1 and 2




In [186]:
!g++ max_pairwise_efficient.cpp -o max_pairwise_efficient -std=c++11
!./max_pairwise_efficient < input1.txt
!./max_pairwise_efficient < input2.txt

90
90


Test and time it with input 3

In [187]:
!g++ max_pairwise_efficient.cpp -o max_pairwise_efficient -std=c++11
!./max_pairwise_efficient < input3.txt

40000000000


20 stress tests

In [188]:
%%writefile stressTestmax_pairwise_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
long long max_pairwise_efficient(const std::vector<int>& numbers) {
    int n = numbers.size(); //1

    long long biggest_1 = std::numeric_limits<long long>::min(); //1
    long long biggest_2 = std::numeric_limits<long long>::min(); //1

    for (int i = 0; i < n; ++i) { //2+2n
        int new_number = numbers[i];
        if (new_number > biggest_1) {
            biggest_2 = biggest_1;
            biggest_1 = new_number;
        } else if (new_number > biggest_2) { //2+2n
            biggest_2 = new_number;
        }
    }

    return biggest_1 * biggest_2;//1
}

int main() {
    // 20 stress tests
    for (int test = 1; test <= 20; ++test) { //2+2n
        int n = rand() % 11 + 2;  // Use a large n
        std::vector<int> numbers(n);

        // Generate n random numbers between 1 and 10^9
        for (int i = 0; i < n; ++i) {        //2+2n
            numbers[i] = rand() % 1000000000 + 1;
        }

        std::cout << "-------------------------------------------------" << std::endl; //1
        std::cout << "Testing with the list: "; //1
        for (int i = 0; i < n; ++i) {           //1
            std::cout << numbers[i] << " ";     //1
        }
        std::cout << std::endl;//1

        std::cout << "The efficient solution gives the result: " << max_pairwise_efficient(numbers) << std::endl; //1

        // Add any additional test or comparison here if needed

        std::cout << "-------------------------------------------------" << std::endl; //1
    }

    return 0;
}




Writing stressTestmax_pairwise_efficient.cpp


In [189]:
!g++ stressTestmax_pairwise_efficient.cpp -o stressTestmax_pairwise_efficient -std=c++11
!./stressTestmax_pairwise_efficient

-------------------------------------------------
Testing with the list: 846930887 681692778 714636916 957747794 424238336 719885387 649760493 596516650 
The efficient solution gives the result: 811146188694713278
-------------------------------------------------
-------------------------------------------------
Testing with the list: 25202363 350490028 783368691 
The efficient solution gives the result: 274562914442913348
-------------------------------------------------
-------------------------------------------------
Testing with the list: 44897764 967513927 365180541 540383427 304089173 
The efficient solution gives the result: 522828491542487829
-------------------------------------------------
-------------------------------------------------
Testing with the list: 35005212 521595369 
The efficient solution gives the result: 18258556470063228
-------------------------------------------------
-------------------------------------------------
Testing with the list: 726956430 33646

Count executed instructions in code and verify with inputs 1,2 and 3

In [190]:
%%writefile max_pairwise_efficient2.cpp
#include <iostream>
#include <vector>
#include <limits>

int execints = 0;

long long max_pairwise_efficient(const std::vector<int>& numbers) {
    int n = numbers.size(); //1
    long long biggest_1 = std::numeric_limits<long long>::min(); //1
    long long biggest_2 = std::numeric_limits<long long>::min(); //1
    execints += 3;

    for (int i = 0; i < n; ++i) { //2+2n
        int new_number = numbers[i];
        long long temp = biggest_1;
        if (new_number > biggest_1) {
            execints += 2;  // Contar la comparación y la actualización
            biggest_2 = biggest_1;
            biggest_1 = new_number;
        }
        if (temp > biggest_2) {
            execints += 2;  // Contar la comparación y la actualización
            biggest_2 = temp;
        }
        execints += 2;  // Contar las dos comparaciones
    }
    execints += 2;  // Contar el aumento final y el retorno

    return execints;

}

int main() {
    int n; //1
    std::cin >> n; //1
    std::vector<int> numbers(n); //1
    for (int i = 0; i < n; ++i) { //2+2n
        std::cin >> numbers[i];
    }
    long long result = max_pairwise_efficient(numbers); //1
    std::cout << "El numero de instrucciones es: " << execints << "\n"; //1
    return 0;
}





Writing max_pairwise_efficient2.cpp


In [191]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_efficient2.cpp -o max_pairwise_efficient2
for i in {1..3}
do
    echo "Input #$i"
    ./max_pairwise_efficient2< input$i.txt
done

Overwriting run_all_test.sh


In [192]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
El numero de instrucciones es: 45
Input #2
El numero de instrucciones es: 29
Input #3
El numero de instrucciones es: 25


Calculate estimated time by instruction and number of instructions by second

In [193]:
time_per_instruction = 0.41/1600007
time_per_instruction

2.5624887891115476e-07

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [194]:
%%writefile max_efficient.py

def max_pairwise_efficient():
    n = int(input())
    values = input().split()
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for new in map(int, values):
        temp = biggest_1
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
        if temp > biggest_2:
            biggest_2 = temp

    return biggest_1 * biggest_2

if __name__ == '__main__':
    print(max_pairwise_efficient())

Writing max_efficient.py


In [195]:
!python3 max_efficient.py


^C


Test with inputs 1 and 2




In [196]:
%%writefile max_efficient2.py

def max_pairwise_efficient():
    n = int(input())
    values = input().split()
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for new in map(int, values):
        temp = biggest_1
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
        elif temp > biggest_2:
            biggest_2 = temp

    return biggest_1 * biggest_2

if __name__ == '__main__':
    print(max_pairwise_efficient())

Writing max_efficient2.py


In [197]:
%%writefile run_all_test2.sh
#comparing all the executions to check the differences between the outputs
for i in {1..2}
do
    python3 max_efficient2.py < input$i.txt > pythonOutput$i.txt
done

Writing run_all_test2.sh


In [198]:
!chmod u+x run_all_test2.sh
!./run_all_test.sh

Input #1
El numero de instrucciones es: 45
Input #2
El numero de instrucciones es: 29
Input #3
El numero de instrucciones es: 25


Test and time it with input 3

In [199]:
!time python3 max_efficient2.py < input3.txt > pythonOutput3.txt


real	0m0.033s
user	0m0.026s
sys	0m0.007s


20 stress tests

In [200]:
#%%writefile stressTestmax_pairwise_efficient.cpp
import random

def max_pairwise_efficient(numbers):
    n = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for i in range(n):
        new = numbers[i]
        temp = biggest_1
        if new > biggest_1:
          biggest_2 = biggest_1
          biggest_1 = new
        if temp > biggest_2: biggest_2 = temp

    return biggest_1 * biggest_2

# 20 stress tests
for _ in range(20):
    n = random.randint(2, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {max_pairwise_efficient(numbers)}")
    # Add any additional test or comparison here if needed
    print("-------------------------------------------------")

-------------------------------------------------
Testing with the list: [787747138, 266606885, 576857338, 284441014, 332349821, 140955674]
The efficient solution gives the result 620545553427191044
-------------------------------------------------
-------------------------------------------------
Testing with the list: [148217723, 71590804, 856668361]
The efficient solution gives the result 126973433833562003
-------------------------------------------------
-------------------------------------------------
Testing with the list: [435843984, 443693628, 618098881, 850499541, 851657723, 33144637]
The efficient solution gives the result 725320877145544729
-------------------------------------------------
-------------------------------------------------
Testing with the list: [42946930, 466767770, 214514107, 475289325, 773129247, 844587387, 289592126, 340908378, 439363840, 294973100, 394907006]
The efficient solution gives the result 713327854279487769
-----------------------------------

Count executed instructions in code and verify with inputs 1,2 and 3

In [201]:
%%writefile max_pairwise_efficient.py
import random
execints = 0

def max_pairwise_efficient(numbers):
    global execints
    n = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')
    execints += 3

    for i in range(n):
        new_number = numbers[i]
        temp = biggest_1
        execints += 2
        if new_number > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new_number
            execints += 2
        if temp > biggest_2:
            biggest_2 = temp
            execints += 2
        execints += 2

    execints += 1
    return biggest_1 * biggest_2

if __name__ == '__main__':
    n = int(input())
    input_line = input()
    numbers = list(map(int, input_line.split()))  # Divide la cadena en números y convierte cada uno a entero
    max_pairwise_efficient(numbers)
    execints += 1

    print("El numero de instrucciones es:", execints)

Writing max_pairwise_efficient.py


In [202]:
%%writefile run_all_test.sh
for i in {1..3}
do
    echo "Input #$i"
    python3 max_pairwise_efficient.py < input$i.txt
done

Overwriting run_all_test.sh


In [203]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
El numero de instrucciones es: 65
Input #2
El numero de instrucciones es: 49
Input #3
El numero de instrucciones es: 37


Calculate estimated time by instruction and number of instructions by second

In [204]:
!time python3 max_pairwise_efficient.py < input11.txt > pythonOutputSort11.txt
!python3 max_pairwise_efficient.py < input11.txt
num_inst_per_seg = 1200005/0.165
print(f"{num_inst_per_seg} tiempo por instruccion")


real	0m0.173s
user	0m0.139s
sys	0m0.030s
El numero de instrucciones es: 1200005
7272757.575757575 tiempo por instruccion


### Estimate and verify the maximun n that can be solved in 1 second

In [205]:
import time
import random
n = 0 #numero de instrucciones

def instruction(numbers):
    m = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')
    global n
    n += 3

    for i in range(m):
        new = numbers[i]
        temp = biggest_1
        n += 2
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
            n += 2
        if temp > biggest_2:
            biggest_2 = temp
            n += 1
        n += 2
    return n
# Ejecutar la función y medir el tiempo de ejecución
start_time = time.time()
numbers = [random.randint(1, 100) for _ in range(10)]
instruction = instruction(numbers) # Suponiendo que 'numbers' es una lista válida
end_time = time.time()

# Calcular el número de instrucciones por segundo
elapsed_time = end_time - start_time
instructions_per_second = instruction / elapsed_time
#print(start_time)
#print(end_time)
print(f"Número de instrucciones por segundo: {instructions_per_second}")


Número de instrucciones por segundo: 16401.1608127924


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

# C++

In [206]:
%%writefile stressTest.cpp

#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

long long max_pairwise_efficient(const std::vector<int>& numbers) { //stressTestC_efficien
    int n = numbers.size(); //1

    long long biggest_1 = std::numeric_limits<long long>::min(); //1
    long long biggest_2 = std::numeric_limits<long long>::min(); //1

    for (int i = 0; i < n; ++i) { //2+2n
        int new_number = numbers[i];
        if (new_number > biggest_1) {
            biggest_2 = biggest_1;
            biggest_1 = new_number;
        } else if (new_number > biggest_2) { //2+2n
            biggest_2 = new_number;
        }
    }

    return biggest_1 * biggest_2;//1
}

long long max_pairwise_product_sort(std::vector<int>& numbers) { //stressTestC_sort
    std::sort(numbers.rbegin(), numbers.rend());
    return static_cast<long long>(numbers[0]) * numbers[1];
}

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());

    for (int i = 0; i < 100; ++i) {
        int n = std::uniform_int_distribution<int>(2, 12)(gen);  // Use a large n
        std::vector<int> numbers(n);
        for (int j = 0; j < n; ++j) {
            numbers[j] = std::uniform_int_distribution<int>(1, 1e9)(gen);  // Generate n random numbers between 1 and 10^9
        }

        std::cout << "-------------------------------------------------\n";
        std::cout << "Testing with the list: ";
        for (int num : numbers) {
            std::cout << num << " ";
        }
        std::cout << "\n";

        long long efficient_result = max_pairwise_efficient(numbers);
        long long comparison_result = max_pairwise_product_sort(numbers);  // Replace with your comparison function result

        std::cout << "The efficient solution gives the result " << efficient_result << "\n";
        std::cout << "The comparison solution gives the result " << comparison_result << "\n";

        if (efficient_result == comparison_result) {
            std::cout << "The solution is correct OK\n";
        } else {
            std::cout << "ERROR\n";
        }

        std::cout << "-------------------------------------------------\n";
    }

    return 0;
}



Overwriting stressTest.cpp


In [207]:
!g++ stressTest.cpp -o stressTest -std=c++11
!./stressTest

-------------------------------------------------
Testing with the list: 277196710 858435415 761848443 743879652 770091049 56517439 888030773 703571377 752562143 739527031 790166784 934951106 
The efficient solution gives the result 830265353378384938
The comparison solution gives the result 830265353378384938
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: 635347824 17529700 380892208 868933996 438643401 639777001 467438578 106849134 848302359 
The efficient solution gives the result 737118758622096564
The comparison solution gives the result 737118758622096564
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: 985481869 453755188 871197392 460958380 
The efficient solution gives the result 858549234136085648
The comparison solution gives the result 858549234136085648
The solution is cor

# Pyhton

In [208]:
import random
def max_pairwise_product_sort(numbers):
  numbers.sort(reverse=True)
  return numbers[0] * numbers[1]

def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]

                if aux_product > max_product:
                  max_product = aux_product

    return max_product


for _ in range(20):  # Run the test 20 times
    n = random.randint(2, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {max_pairwise_product_sort(numbers)}")
    print(f"The comparison solution gives the result {max_pairwise_product(numbers)}")
    if max_pairwise_product_sort(numbers) == max_pairwise_product(numbers):
        print("The solution is correct OK")
    else:
        print("ERROR")
    print("-------------------------------------------------")

-------------------------------------------------
Testing with the list: [384738451, 573552852, 120202792, 550254399, 891745293, 284632998, 239423350]
The efficient solution gives the result 511463056057725636
The comparison solution gives the result 511463056057725636
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [930087999, 634410263, 633611121, 525193610, 242773998]
The efficient solution gives the result 590057372058733737
The comparison solution gives the result 590057372058733737
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [296699866, 955137409, 414159896]
The efficient solution gives the result 395579609977149464
The comparison solution gives the result 395579609977149464
The solution is correct OK
-------------------------------------------------
-------------------------